internal/common/client.go

```diff
@@ -1,44 +1,117 @@
-package common
-
-import "fmt"
-
-// ClientID is an enum for client IDs
-type ClientID int
-
-const (
-	// Team1 ID
-	Team1 ClientID = iota
-	// Team2 ID
-	Team2
-	// Team3 ID
-	Team3
-	// Team4 ID
-	Team4
-	// Team5 ID
-	Team5
-	// Team6 ID
-	Team6
-)
-
-// TeamIDs contain sequential IDs of all teams
-var TeamIDs = [...]ClientID{Team1, Team2, Team3, Team4, Team5, Team6}
-
-// Client is a base interface to be implemented by each client struct.
-type Client interface {
-	Echo(s string) string
-	GetID() ClientID
-	Logf(format string, a ...interface{})
-}
-
-// RegisteredClients contain all registered clients, exposed for the server.
-var RegisteredClients = map[ClientID]Client{}
-
-// RegisterClient registers clients into RegisteredClients
-func RegisterClient(id ClientID, c Client) {
-	// prevent double registrations
-	if _, ok := RegisteredClients[id]; ok {
-		// OK to panic here, as this is a _crucial_ step.
-		panic(fmt.Sprintf("Duplicate client ID %v in RegisterClient!", id))
-	}
-	RegisteredClients[id] = c
-}
+package common
+
+import (
+	"errors"
+	"fmt"
+	"github.com/SOMAS2020/SOMAS2020/internal/common/rules"
+	"gonum.org/v1/gonum/mat"
+	"strconv"
+)
+
+// ClientID is an enum for client IDs
+type ClientID int
+
+const (
+	// Team1 ID
+	Team1 ClientID = iota
+	// Team2 ID
+	Team2
+	// Team3 ID
+	Team3
+	// Team4 ID
+	Team4
+	// Team5 ID
+	Team5
+	// Team6 ID
+	Team6
+)
+
+// TeamIDs contain sequential IDs of all teams
+var TeamIDs = [...]ClientID{Team1, Team2, Team3, Team4, Team5, Team6}
+
+// Client is a base interface to be implemented by each client struct.
+type Client interface {
+	Echo(s string) string
+	GetID() ClientID
+	Logf(format string, a ...interface{})
+}
+
+// RegisteredClients contain all registered clients, exposed for the server.
+var RegisteredClients = map[ClientID]Client{}
+
+// RegisterClient registers clients into RegisteredClients
+func RegisterClient(id ClientID, c Client) {
+	// prevent double registrations
+	if _, ok := RegisteredClients[id]; ok {
+		// OK to panic here, as this is a _crucial_ step.
+		panic(fmt.Sprintf("Duplicate client ID %v in RegisterClient!", id))
+	}
+	RegisteredClients[id] = c
+}
+
+// BasicRuleEvaluator implements a basic version of the Matrix rule evaluator, provides single boolean output (and error if present)
+func BasicRuleEvaluator(ruleName string) (bool, error) {
+	if rm, ok := rules.AvailableRules[ruleName]; ok {
+		variables := rm.RequiredVariables
+
+		var variableVect []float64
```

Comment:
move this above L54 so we can immediately see the for loop is iterating through `rm.RequiredVariables`?

Reply:
Fair enough, thanks